cmake/initialize-submodules.cmake

```diff
@@ -1,27 +1,89 @@
+#update the named submodule
+
+#attempt to add the submodule
+#optional arguments:
+# 1 - path
+# 2 - alternate reference name
+macro(add_submodule module_name)
+
+  message(STATUS "Attempting to add \"${module_name}\"")
+  #check of specific path is provided
+  set(MacroArgs ${ARGN})
+  list(LENGTH MacroArgs NumArgs)
+
+  set(have_submodule FALSE)
+  set(have_path FALSE)
+
+  # first optional argument is module path
+  if(NumArgs GREATER 0)
+    set(module_path ${ARGV1})
+  else()
+    set(module_path ${CMAKE_CURRENT_SOURCE_DIR}/${module_name})
+  endif()
+
+  # second optional argument is module reference
+  if(NumArgs GREATER 1)
+    set(module_ref_name ${ARGV2})
+  else()
+    set(module_ref_name ${module_name})
+  endif()
+
+  set(is_submodule FALSE)
   if(GIT_FOUND AND EXISTS "${PROJECT_SOURCE_DIR}/.git")
     set(git_modules_file "${PROJECT_SOURCE_DIR}/.gitmodules")
     if (EXISTS ${git_modules_file})
       file(STRINGS ${git_modules_file} file_lines)
+      set(have_path FALSE)
+
       foreach(line ${file_lines})
-        if (${line} MATCHES "url =")
-          string(REGEX REPLACE "\\s*url = .*/(.*).git" "\\1" submodule "${line}")
-          string(STRIP "${submodule}" submodule)
-          if(NOT EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/${submodule}")
-            message(FATAL_ERROR "Submodule directory \"${CMAKE_CURRENT_SOURCE_DIR}/${submodule}\" does not exist")
+        if (NOT is_submodule)
+          string(COMPARE EQUAL ${line} "[submodule \"${module_ref_name}\"]" is_submodule)
+          if (is_submodule)
+            message(STATUS "\"${module_name}\" is a submodule of this project.")
+            continue()
           endif()
-          # Initialize submodule if it hasn't already been cloned
-          if(NOT EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/${submodule}/.git")
-            message(STATUS "Initialize ${submodule} submodule")
-            execute_process(COMMAND ${GIT_EXECUTABLE} submodule update --init --recursive "${CMAKE_CURRENT_SOURCE_DIR}/${submodule}"
-              WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
-              RESULT_VARIABLE GIT_SUBMOD_RESULT)
-            if(NOT GIT_SUBMOD_RESULT EQUAL "0")
-              message(FATAL_ERROR "git submodule update --init --recursive ${CMAKE_CURRENT_SOURCE_DIR}/${submodule} failed with ${GIT_SUBMOD_RESULT}, please checkout submodules")
-            endif()
+        endif()
+
+        if (is_submodule AND (NOT have_path))
+          string(FIND ${line} "path =" pos)
+          set(have_path NOT(pos EQUALS -1))
+          if (have_path)
+            string(REPLACE "path =" "" submodule_path ${line})
+            string(STRIP "${submodule_path}" submodule_path)
+            set(submodule_path "${PROJECT_SOURCE_DIR}/${submodule_path}")
+            continue()
           endif()
         endif()
       endforeach()
+
+      if(is_submodule)
+        if(have_path)
+          message(STATUS "Updating submodule \"${module_name}\" at ${submodule_path}")
```

Comment:
I think this is where we want a status message since cloning can take a while and it explains what's happening on stdout. Generally, this should only happen for the initial clone of the repository (and not for updates). Maybe a better message would be: "Initializing ${submodule} submodule"?